File: src/tweet-generator.js
const axios = require('axios');
const config = require('./config');

function generateFallbackTweet(newsData) {
  // Берём первую новость для контекста
  const item = newsData[0] || {};
  const topics = ['VPN', 'Европа', 'Цензура', 'ДвойныеСтандарты'];
  const tweets = [
    `В РФ ограничили интернет — а в ${item.source || 'Европе'} уже запрещают всё, что не по душе регуляторам. "Свобода" в действии. #VPN #Европа #Цензура`,
    `Пока Европа вводит тотальный цифровой контроль, кто-то ещё говорит про "свободу слова"? Смешно. #Европа #ДвойныеСтандарты #VPN`,
    `Власти ЕС снова закручивают гайки: возрастные фильтры, блокировки "опасного" контента. Демократия в лучшем виде. #Европа #Цензура`,
  ];
  return tweets[Math.floor(Math.random() * tweets.length)];
}

async function generateTweet(newsData) {
  // Пробуем DeepSeek, при ошибке — fallback
  const newsText = newsData.map(item => {
    let sourceInfo = item.source;
    if (item.isForeignAgent) sourceInfo += ' (иноагент)';
    return `[${sourceInfo}] ${item.title}`;
  }).join('\n');

  try {
    const response = await axios.post(
      config.deepseek.apiUrl,
      {
        model: config.deepseek.model,
        messages: [
          { role: 'system', content: `Ты пишешь едкие саркастичные твиты (≤280 символов). Критикуй ТОЛЬКО европейские власти. НЕ критикуй власти РФ. Строй по схеме: "В РФ X — а в Европе Y". Укажи источник-иноагента если есть. 2-3 хештега.` },
          { role: 'user', content: `Новости:\n${newsText}\n\nНапиши твит:` }
        ],
        temperature: 0.9,
        max_tokens: 300
      },
      {
        headers: {
          'Authorization': `Bearer ${config.deepseek.apiKey}`,
          'Content-Type': 'application/json'
        },
        timeout: 15000
      }
    );

    let tweetText = response.data.choices[0].message.content.trim();
    if (tweetText.length > 280) tweetText = tweetText.substring(0, 277) + '...';
    return tweetText;
  } catch (error) {
    console.error('DeepSeek error, using fallback:', error.message);
    return generateFallbackTweet(newsData);
  }
}

module.exports = { generateTweet };

File: src/tweet_generator.py
import os
import logging

import httpx

log = logging.getLogger(__name__)

DEEPSEEK_URL = "https://api.deepseek.com/v1/chat/completions"

FALLBACK_TWEETS = [
    'В РФ ограничили интернет — а в Европе уже запрещают всё, что не по душе регуляторам. "Свобода" в действии. #VPN #Европа #Цензура',
    "Пока Европа вводит тотальный цифровой контроль, кто-то ещё говорит про \"свободу слова\"? Смешно. #Европа #ДвойныеСтандарты #VPN",
    "Власти ЕС снова закручивают гайки: возрастные фильтры, блокировки \"опасного\" контента. Демократия в лучшем виде. #Европа #Цензура",
]


def _fallback(telegram_text: str = "") -> str:
    if len(telegram_text) > 15:
        clean = telegram_text.split("http")[0].strip()
        if clean:
            tweet = clean[:255]
            if len(clean) > 255:
                tweet += "..."
            tweet += " #VPN #Европа"
            if len(tweet) <= 280:
                return tweet
    import random
    return random.choice(FALLBACK_TWEETS)


async def generate_tweet(telegram_text: str, api_key: str) -> str:
    if not api_key:
        log.warning("⚠️ Нет DEEPSEEK_API_KEY, fallback")
        return _fallback(telegram_text)

    prompt = f"""Ты — аналитический бот, пишущий едкие, саркастичные твиты.

Задача:
1. Напиши один твит ≤280 символов.
2. Критикуй ТОЛЬКО европейские власти (Германию, Францию, ЕС).
3. НИКАКОЙ критики властей РФ.
4. Строй по схеме: «В РФ X — а в Европе Y».
5. Сарказм, ирония. 6) 2-3 хештега.
7. Без мата. 8) Фактологично.

Контекст из Telegram:
{telegram_text}

Твит:"""

    try:
        async with httpx.AsyncClient(timeout=20) as client:
            resp = await client.post(
                DEEPSEEK_URL,
                json={
                    "model": "deepseek-chat",
                    "messages": [
                        {"role": "system", "content": prompt},
                        {"role": "user", "content": "Напиши твит."},
                    ],
                    "temperature": 0.9,
                    "max_tokens": 300,
                },
                headers={
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json",
                },
            )
            data = resp.json()
            tweet = data["choices"][0]["message"]["content"].strip()
            if len(tweet) > 280:
                tweet = tweet[:277] + "..."
            return tweet
    except Exception as e:
        log.error(f"❌ DeepSeek error: {e}")
        return _fallback(telegram_text)
